### ISSUE TITLE
[bug] Device dispatcher can choose CPU path for CUDA tensors.

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#79202]
### 🐛 Describe the bug

It is possible to confuse device dispatcher to choose wrong path when device dispatches are specified in a single line in `native_functions.yaml`.

To reproduce checkout the code from https://github.com/pytorch/pytorch/pull/79201 at https://github.com/pytorch/pytorch/pull/79201/commits/d03cfcfa406a75a5bcd11e0f6adeaa461b338e07.
Then
```python
In [1]: import torch

In [2]: ig = torch.rand(3, 3, device='cuda')

In [3]: ic = ig.to('cpu')

In [4]: res = torch.foo(ic, ic, ic)
CPU kernel

In [5]: res = torch.foo(ig, ig, ig)
CPU kernel
Segmentation fault (core dumped)

```

### Versions

master

**Comment by nikitaved** [comment]
Right, and there are functions that have `CPU, CUDA` dispatches specified in a single line with the same name. `angle`, for example.

**Comment by xsacha** [comment]
Can't you have a static build in CI to trigger the duplicate symbol error?

**Comment by ezyang** [comment]
I guess... not sure if the CI team would want to maintain it though
